test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/TestFhirServer.cs

```diff
@@ -52,26 +69,125 @@ public TestFhirClient GetTestFhirClient(ResourceFormat format, TestApplication c
 
             if (!reusable)
             {
-                return CreateFhirClient(format, clientApplication, user);
+                return CreateFhirClient(format, clientApplication, user, authenticationHandler);
             }
 
             return _cache.GetOrAdd(
                     (format, clientApplication, user),
                     (tuple, fhirServer) =>
-                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user)),
+                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user, authenticationHandler)),
                     this)
                 .Value;
         }
 
-        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user)
+        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user, DelegatingHandler authenticationHandler = null)
         {
-            var httpClient = new HttpClient(new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer)) { BaseAddress = BaseAddress };
+            ConfigureSecurityOptions().GetAwaiter().GetResult();
+
+            var sessionMessageHandler = new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer);
+
+            if (authenticationHandler != null)
+            {
+                authenticationHandler.InnerHandler = CreateMessageHandler();
+                sessionMessageHandler.InnerHandler = authenticationHandler;
+            }
+            else if (SecurityEnabled && !clientApplication.Equals(TestApplications.InvalidClient))
+            {
+                string authKey = GenerateKey();
+                if (_authenticationHandlers.ContainsKey(authKey))
+                {
+                    sessionMessageHandler.InnerHandler = _authenticationHandlers[authKey];
+                }
+                else
+                {
+                    AuthenticationHttpMessageHandler authHandler;
+
+                    var authHttpClient = new HttpClient(new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer)) { BaseAddress = BaseAddress };
```

Comment:
If we do this in the client handler, then we dont need to create new HttpClient each time. The SendAsync could just execute additional HttpRequestMessage (to get token and etc) before execute the HttpRequestMessge that was passed in?

Reply:
The `HttpClient` is used in the implementations of the `ICredentialProvider` that we're using here. There are additional `ICredentialProviders` that don't directly make `HttpClient` calls such as `ManagedIdentityCredentialProvider`. All the `AuthenticationHttpMessageHandler` knows is that it has an `ICredentialProvider` that exposes a method `GetBearerToken()`. 

Again, I think this may be a bit of a test specific thing in that in most cases you wouldn't be creating multiple `ICredentialProviders`, but only 1 of them for a life of an application. In this case, we're creating one per cached handler.